cmd/mrf.go

```diff
@@ -15,51 +15,199 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+//go:generate msgp -file=$GOFILE
+
 package cmd
 
 import (
 	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/google/uuid"
 	"github.com/minio/madmin-go/v3"
 	"github.com/minio/pkg/v3/wildcard"
+	"github.com/tinylib/msgp/msgp"
 )
 
 const (
 	mrfOpsQueueSize = 100000
 )
 
-// partialOperation is a successful upload/delete of an object
+const (
+	healDir              = ".heal"
+	healMRFDir           = bucketMetaPrefix + SlashSeparator + healDir + SlashSeparator + "mrf"
+	healMRFMetaFormat    = 1
+	healMRFMetaVersionV1 = 1
+)
+
+// PartialOperation is a successful upload/delete of an object
 // but not written in all disks (having quorum)
-type partialOperation struct {
-	bucket              string
-	object              string
-	versionID           string
-	versions            []byte
-	setIndex, poolIndex int
-	queued              time.Time
-	scanMode            madmin.HealScanMode
+type PartialOperation struct {
+	Bucket              string
+	Object              string
+	VersionID           string
+	Versions            []byte
+	SetIndex, PoolIndex int
+	Queued              time.Time
+	BitrotScan          bool
 }
 
 // mrfState sncapsulates all the information
 // related to the global background MRF.
 type mrfState struct {
-	opCh chan partialOperation
+	opCh chan PartialOperation
+
+	closing int32
+	wg      sync.WaitGroup
+}
+
+func newMRFState() mrfState {
+	return mrfState{
+		opCh: make(chan PartialOperation, mrfOpsQueueSize),
+	}
 }
 
 // Add a partial S3 operation (put/delete) when one or more disks are offline.
-func (m *mrfState) addPartialOp(op partialOperation) {
+func (m *mrfState) addPartialOp(op PartialOperation) {
 	if m == nil {
 		return
 	}
 
+	m.wg.Add(1)
+	defer m.wg.Done()
+
+	if atomic.LoadInt32(&m.closing) == 1 {
+		return
+	}
+
 	select {
 	case m.opCh <- op:
 	default:
 	}
 }
 
+// Do not accept new MRF operations anymore and start to save
+// the current heal status in one available disk
+func (m *mrfState) shutdown() {
+	atomic.StoreInt32(&m.closing, 1)
+
+	m.wg.Wait()
```

Comment:
The wait group approach to synchronize between the multiple senders, i.e `addPartialOp` and the `shutdown` method won't work since all calls to `wg.Add` must happen before we call `wg.Wait`. This isn't guaranteed.
> Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.
ref: https://pkg.go.dev/sync#WaitGroup.Add

I understand that closing `m.opCh` simplifies processing of outstanding `PartialOp` values onto local drives. But closing a channel which has multiple senders is error prone and complicates the code. 

As an alternate approach, we could use a done/quit channel, which will be closed at the beginning of the `shutdown` method. `addPartialOp` calls will select on this channel and return immediately if they found it closed.
This way we stop more tasks from being sent on `m.opCh`.  Now, we could drain at most `len(m.opCh)` entries from  `m.opCh`. Of course this is only a point in time value of the number of outstanding `PartialOp` values and there could be more sent after we signalled on the quit channel (see above). The downside is that we may drop the 'last' few values. Overall, the implementation will be easier to reason with.

Reply:
> The wait group approach to synchronize between the multiple senders, i.e addPartialOp and the shutdown method won't work since all calls to wg.Add must happen before we call wg.Wait. This isn't guaranteed.

I do not need a guarantee here. I just wanted to avoid sending new entries to m.opCh during shutdown